torch/functional.py

```diff
@@ -1649,3 +1651,62 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
 
 def align_tensors(*tensors):
     raise RuntimeError('`align_tensors` not yet implemented.')
+
+def unravel_index(
+    indices: Tensor,
+    shape: Union[Tensor, Tuple[int, ...]],
+    *,
+    as_tuple: bool = True
+) -> Union[Tuple[Tensor, ...], Tensor]:
+    r"""Converts flat indices into unraveled coordinates for the given target shape.
+    Args:
+        indices: A `Tensor` containing flattened indices to be unraveled, should be integral type.
+        shape: The shape (can be a `tuple`, `list` or a `Tensor`) used for unravelling the `indices`.
+    Keyword Args:
+        as_tuple: A boolean value, which if `True` will return the unraveled `indices` as tuple,
+                  else a `Tensor` will be returned. Default: `False`
+    Returns:
+        unraveled coordinates from the given `indices`, and `shape`. See description of `as_tuple` for
+        returning a tuple.
+    Example::
+        >>> indices = torch.tensor([22, 41, 37])
+        >>> shape = (7, 6)
+        >>> torch.unravel_index(indices, shape)
+        tensor([[3, 4],
+                [6, 5],
+                [6, 1]])
+        >>> torch.unravel_index(indices, shape, as_tuple=True)
+        (tensor([3, 6, 6]), tensor([4, 5, 1]))
+    """
+    def _helper_type_check(tens: Union[Sequence, Tensor], name: str):
+        if isinstance(tens, Tensor):
+            if tens.dtype not in integral_types():
+                raise TypeError(f"Unsupported dtype for shape {tens.dtype}, only integral types allowed.")
+            if torch.any(tens < 0):
+                raise ValueError(f"Negative values in {name} are not allowed")
+        else:
+            raise TypeError(f"{name} should either be a Sequence or a Tensor, but found {type(tens)}.")
+
+    _helper_type_check(indices, "indices")
+
+    if isinstance(shape, Sequence):
+        # Convert to a tensor, with the same properties as that of indices
+        shape = indices.new_tensor(shape)
+    _helper_type_check(shape, "shape")
+
+    # In case no indices passed, short-circuit to return itself
+    if indices.numel() == 0:
+        return tuple(indices) if as_tuple else indices
+
+    if shape.numel() == 0:
+        raise ValueError(f"Empty shape tensor passed, expected shape tensor of minimum size {indices.shape}.")
+
+    shape = torch.cat((shape, torch.tensor((1,))), dim=0)
+    if torch.max(indices) >= torch.prod(shape):
+        raise ValueError("Given indices should be representible for the given shape.")
+
+    coefs = shape[1:].flipud().cumprod(dim=0).flipud()
+    coords = torch.div(indices[..., None], coefs, rounding_mode='trunc') % shape[:-1]
+    if as_tuple:
+        return tuple(coords[..., i] for i in range(coords.size(-1)))
+    return coords
```

Comment:
I think concatenating `shape` with `torch.tensor((1,))` throws an error if `shape` is on CUDA. Using `shape.new_tensor((1,))` should fix the problem. Also, we can append 1 to `coefs`, instead of `shape`, which is slightly easier to understand.

```suggestion
    if torch.max(indices) >= torch.prod(shape):
        raise ValueError("Given indices should be representible for the given shape.")

    coefs = torch.cat((shape[1:], shape.new_tensor((1,))), dim=0).flipud().cumprod(dim=0).flipud()
    coords = torch.div(indices[..., None], coefs, rounding_mode='trunc') % shape

    if as_tuple:
        return tuple(coords[..., i] for i in range(coords.size(-1)))
    return coords
```

Reply:
Thanks, @francois-rozet for the suggestion. This code can be further cleaned, will send the changes by tomorrow, (the tests need updates as well). :) Appreciate the review.